File: app/app/src/main/java/com/shelter/shade/engine/Raycast.kt
package com.shelter.shade.engine

import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.tan

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

private const val EPS = 1e-9

// ---------------------------------------------------------------------------
// Data models
// ---------------------------------------------------------------------------

/**
 * Building footprint projected onto the local ENU (East/North) xy plane.
 *
 * [ringXy] elements are doubleArrayOf(x, y) where x = east metres, y = north metres.
 */
data class ProjectedBuilding(
    val ringXy: List<DoubleArray>,
    val heightM: Double,
    val heightEstimated: Boolean = false,
    val osmId: String? = null,
) {
    /** Returns [minX, minY, maxX, maxY]. */
    fun bbox(): DoubleArray {
        var minX = Double.MAX_VALUE
        var minY = Double.MAX_VALUE
        var maxX = -Double.MAX_VALUE
        var maxY = -Double.MAX_VALUE
        for (pt in ringXy) {
            if (pt[0] < minX) minX = pt[0]
            if (pt[1] < minY) minY = pt[1]
            if (pt[0] > maxX) maxX = pt[0]
            if (pt[1] > maxY) maxY = pt[1]
        }
        return doubleArrayOf(minX, minY, maxX, maxY)
    }
}

/**
 * Shade determination result for a single point.
 *
 * [reason] is one of: "sunny", "building", "inside_building", "sun_below".
 * [confidence] is in [0, 1]; reduced when building height is estimated and
 * close to the threshold.
 */
data class ShadeResult(
    val shaded: Boolean,
    val reason: String,
    val confidence: Double,
    val blockerId: String? = null,
    val blockerDistanceM: Double? = null,
)

// ---------------------------------------------------------------------------
// Geometry helpers (package-private — used by tests via same package)
// ---------------------------------------------------------------------------

/**
 * Point-in-polygon test using the ray casting algorithm.
 *
 * Points on the boundary are NOT considered inside (approximate).
 * Mirrors Python `_point_in_ring`.
 */
internal fun pointInRing(px: Double, py: Double, ring: List<DoubleArray>): Boolean {
    var inside = false
    val n = ring.size
    var j = n - 1
    for (i in 0 until n) {
        val xi = ring[i][0]; val yi = ring[i][1]
        val xj = ring[j][0]; val yj = ring[j][1]
        if ((yi > py) != (yj > py)) {
            val xCross = xi + (py - yi) / (yj - yi) * (xj - xi)
            if (px < xCross) inside = !inside
        }
        j = i
    }
    return inside
}

/**
 * Intersection distance between a ray (origin o, unit direction d) and a
 * line segment a→b.
 *
 * Returns the distance t ≥ 0 along the ray, or null if there is no
 * intersection (parallel or behind the origin or outside the segment).
 *
 * The derivation uses the 2-D cross-product formulation:
 *
 *   v1 = o - a
 *   v2 = b - a
 *   v3 = perp(d) = (-dy, dx)
 *
 *   t1 = cross(v2, v1) / dot(v2, v3)   [ray parameter]
 *   t2 = dot(v1, v3)  / dot(v2, v3)    [segment parameter ∈ [0,1]]
 *
 * This is identical to the Python `_ray_segment_distance` implementation.
 */
internal fun raySegmentDistance(
    ox: Double, oy: Double,
    dx: Double, dy: Double,
    ax: Double, ay: Double,
    bx: Double, by: Double,
): Double? {
    val v1x = ox - ax;  val v1y = oy - ay
    val v2x = bx - ax;  val v2y = by - ay
    val v3x = -dy;      val v3y = dx
    val denom = v2x * v3x + v2y * v3y
    if (abs(denom) < EPS) return null        // parallel
    val t1 = (v2x * v1y - v2y * v1x) / denom   // ray distance
    val t2 = (v1x * v3x + v1y * v3y) / denom   // segment parameter
    return if (t1 >= 0.0 && t2 >= 0.0 && t2 <= 1.0) t1 else null
}

/**
 * Minimum intersection distance between a ray and all edges of a polygon ring.
 *
 * Returns null when the ray does not intersect any edge.
 * Mirrors Python `_ray_ring_min_distance`.
 */
internal fun rayRingMinDistance(
    ox: Double, oy: Double,
    dx: Double, dy: Double,
    ring: List<DoubleArray>,
): Double? {
    var best: Double? = null
    val n = ring.size
    for (i in 0 until n) {
        val ax = ring[i][0];         val ay = ring[i][1]
        val bx = ring[(i + 1) % n][0]; val by = ring[(i + 1) % n][1]
        val d = raySegmentDistance(ox, oy, dx, dy, ax, ay, bx, by)
        if (d != null && (best == null || d < best)) best = d
    }
    return best
}

// ---------------------------------------------------------------------------
// Core shade determination
// ---------------------------------------------------------------------------

/**
 * Determine whether [pointXy] is shaded at the given solar position.
 *
 * [pointXy]       — [x, y] in local ENU metres (x = east, y = north).
 * [sunAzimuthDeg] — clockwise from north, degrees.
 * [sunAltitudeDeg]— elevation above horizon, degrees.
 * [buildings]     — projected building footprints to test against.
 * [maxDistanceM]  — ignore buildings further than this (metres).
 * [minAltitudeDeg]— sun altitude at or below this → "sun_below".
 *
 * Mirrors Python `is_point_shaded` exactly.
 */
fun isPointShaded(
    pointXy: DoubleArray,
    sunAzimuthDeg: Double,
    sunAltitudeDeg: Double,
    buildings: List<ProjectedBuilding>,
    maxDistanceM: Double = 1000.0,
    minAltitudeDeg: Double = 0.5,
): ShadeResult {
    if (sunAltitudeDeg <= minAltitudeDeg) {
        return ShadeResult(shaded = true, reason = "sun_below", confidence = 1.0)
    }

    val ox = pointXy[0]; val oy = pointXy[1]
    val az = Math.toRadians(sunAzimuthDeg)
    val dx = sin(az); val dy = cos(az)          // unit vector toward sun (east, north)
    val tanAlt = tan(Math.toRadians(sunAltitudeDeg))

    var nearestBlockerDist: Double? = null
    var nearestBlockerId: String? = null
    var confidence = 1.0

    for (b in buildings) {
        val bb = b.bbox()
        val minX = bb[0]; val minY = bb[1]; val maxX = bb[2]; val maxY = bb[3]
        // Fast bbox pre-filter: corridor distance lower-bound
        val ddx = max(minX - ox, max(0.0, ox - maxX))
        val ddy = max(minY - oy, max(0.0, oy - maxY))
        if (hypot(ddx, ddy) > maxDistanceM) continue

        if (pointInRing(ox, oy, b.ringXy)) {
            return ShadeResult(
                shaded = true,
                reason = "inside_building",
                confidence = if (!b.heightEstimated) 1.0 else 0.7,
                blockerId = b.osmId,
                blockerDistanceM = 0.0,
            )
        }

        val d = rayRingMinDistance(ox, oy, dx, dy, b.ringXy) ?: continue
        if (d > maxDistanceM) continue

        val required = d * tanAlt    // height needed to block sun at this distance
        if (b.heightM >= required) {
            if (nearestBlockerDist == null || d < nearestBlockerDist) {
                nearestBlockerDist = d
                nearestBlockerId = b.osmId
            }
            // Confidence reduction: estimated height AND close to threshold
            if (required > 0 && b.heightEstimated && b.heightM < required * 1.2) {
                confidence = min(confidence, 0.6)
            }
        }
    }

    return if (nearestBlockerDist != null) {
        ShadeResult(
            shaded = true,
            reason = "building",
            confidence = confidence,
            blockerId = nearestBlockerId,
            blockerDistanceM = nearestBlockerDist,
        )
    } else {
        ShadeResult(shaded = false, reason = "sunny", confidence = confidence)
    }
}

// ---------------------------------------------------------------------------
// Spatial index
// ---------------------------------------------------------------------------

/**
 * Uniform-grid spatial index over projected buildings.
 *
 * Each building is bucketed into all grid cells its bounding box overlaps.
 * [candidates] returns a deduplicated list of buildings in all cells that
 * could contain buildings within [radius] of (x, y).
 *
 * Mirrors Python `BuildingIndex`.
 */
class BuildingIndex(buildings: List<ProjectedBuilding>, private val cellSizeM: Double = 64.0) {

    // grid cell → list of buildings whose bbox overlaps that cell
    private val grid: HashMap<Long, MutableList<ProjectedBuilding>> = HashMap()

    init {
        for (b in buildings) {
            val bb = b.bbox()
            val minX = bb[0]; val minY = bb[1]; val maxX = bb[2]; val maxY = bb[3]
            for (cx in cell(minX)..cell(maxX)) {
                for (cy in cell(minY)..cell(maxY)) {
                    grid.getOrPut(packCell(cx, cy)) { mutableListOf() }.add(b)
                }
            }
        }
    }

    private fun cell(v: Double): Int = floor(v / cellSizeM).toInt()

    /** Pack two ints into a Long for use as a HashMap key. */
    private fun packCell(cx: Int, cy: Int): Long =
        (cx.toLong() and 0xFFFF_FFFFL) or (cy.toLong() shl 32)

    /**
     * Return all unique candidate buildings within cells that could contain
     * buildings within [radius] metres of ([x], [y]).
     *
     * Uses object identity (System.identityHashCode) to deduplicate, mirroring
     * Python's `id(b)` dict key.
     */
    fun candidates(x: Double, y: Double, radius: Double): List<ProjectedBuilding> {
        val seen = LinkedHashMap<Int, ProjectedBuilding>()
        for (cx in cell(x - radius)..cell(x + radius)) {
            for (cy in cell(y - radius)..cell(y + radius)) {
                val bucket = grid[packCell(cx, cy)] ?: continue
                for (b in bucket) {
                    seen[System.identityHashCode(b)] = b
                }
            }
        }
        return seen.values.toList()
    }

    /**
     * Shade determination using the spatial index.
     *
     * Mirrors Python `BuildingIndex.is_point_shaded`.
     */
    fun isPointShaded(
        pointXy: DoubleArray,
        sunAzimuthDeg: Double,
        sunAltitudeDeg: Double,
        maxDistanceM: Double = 1000.0,
        minAltitudeDeg: Double = 0.5,
    ): ShadeResult {
        if (sunAltitudeDeg <= minAltitudeDeg) {
            return ShadeResult(shaded = true, reason = "sun_below", confidence = 1.0)
        }
        val cands = candidates(pointXy[0], pointXy[1], maxDistanceM)
        return isPointShaded(pointXy, sunAzimuthDeg, sunAltitudeDeg, cands, maxDistanceM, minAltitudeDeg)
    }
}

File: shade-engine/shade_engine/raycast.py
"""레이캐스팅 그늘 판정 (엔진의 심장).

한 지점이 그늘인지 판정: 태양 방향(수평 방위각)으로 광선을 쏴, 그 경로상의
건물이 (태양 고도각으로 결정되는) 그림자 임계 높이보다 높으면 그늘이다.
지면 관측자(높이 0), 평지붕(flat roof) 가정의 1차 근사.

좌표는 로컬 ENU 미터 평면(x=동, y=북)에서 처리한다. 위경도 → xy 변환은 engine
계층에서 수행한다.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

_EPS = 1e-9


@dataclass(frozen=True)
class ProjectedBuilding:
    """로컬 xy 평면으로 투영된 건물."""

    ring_xy: tuple[tuple[float, float], ...]
    height_m: float
    height_estimated: bool = False
    osm_id: str | None = None

    def bbox(self) -> tuple[float, float, float, float]:
        xs = [p[0] for p in self.ring_xy]
        ys = [p[1] for p in self.ring_xy]
        return (min(xs), min(ys), max(xs), max(ys))


@dataclass(frozen=True)
class ShadeResult:
    """한 지점의 그늘 판정 결과."""

    shaded: bool
    reason: str  # 'sunny' | 'building' | 'inside_building' | 'sun_below'
    confidence: float  # 0~1, 건물 높이 추정/경계 근접 시 하락
    blocker_id: str | None = None
    blocker_distance_m: float | None = None


def _point_in_ring(px: float, py: float, ring: tuple[tuple[float, float], ...]) -> bool:
    """다각형 내부 판정(ray casting). 경계 위는 내부로 간주하지 않는다(근사)."""
    inside = False
    n = len(ring)
    j = n - 1
    for i in range(n):
        xi, yi = ring[i]
        xj, yj = ring[j]
        if (yi > py) != (yj > py):
            x_cross = xi + (py - yi) / (yj - yi) * (xj - xi)
            if px < x_cross:
                inside = not inside
        j = i
    return inside


def _ray_segment_distance(
    ox: float, oy: float, dx: float, dy: float, ax: float, ay: float, bx: float, by: float
) -> float | None:
    """광선(원점 o, 단위방향 d)과 선분(a→b)의 교차 거리. 없으면 None."""
    v1x, v1y = ox - ax, oy - ay
    v2x, v2y = bx - ax, by - ay
    v3x, v3y = -dy, dx
    denom = v2x * v3x + v2y * v3y
    if abs(denom) < _EPS:
        return None  # 평행
    t1 = (v2x * v1y - v2y * v1x) / denom  # 광선 거리
    t2 = (v1x * v3x + v1y * v3y) / denom  # 선분 파라미터
    if t1 >= 0.0 and 0.0 <= t2 <= 1.0:
        return t1
    return None


def _ray_ring_min_distance(
    ox: float, oy: float, dx: float, dy: float, ring: tuple[tuple[float, float], ...]
) -> float | None:
    """광선과 폴리곤 외곽(모든 변)의 최단 교차 거리."""
    best: float | None = None
    n = len(ring)
    for i in range(n):
        ax, ay = ring[i]
        bx, by = ring[(i + 1) % n]
        d = _ray_segment_distance(ox, oy, dx, dy, ax, ay, bx, by)
        if d is not None and (best is None or d < best):
            best = d
    return best


def is_point_shaded(
    point_xy: tuple[float, float],
    sun_azimuth_deg: float,
    sun_altitude_deg: float,
    buildings: list[ProjectedBuilding],
    *,
    max_distance_m: float = 1000.0,
    min_altitude_deg: float = 0.5,
) -> ShadeResult:
    """한 지점이 그늘인지 판정한다.

    sun_altitude_deg 가 임계치 이하면 직사광선이 의미 없으므로 그늘로 본다.
    """
    if sun_altitude_deg <= min_altitude_deg:
        return ShadeResult(shaded=True, reason="sun_below", confidence=1.0)

    ox, oy = point_xy
    az = math.radians(sun_azimuth_deg)
    dx, dy = math.sin(az), math.cos(az)  # 태양을 향하는 수평 단위벡터(동,북)
    tan_alt = math.tan(math.radians(sun_altitude_deg))

    nearest_blocker_dist: float | None = None
    nearest_blocker_id: str | None = None
    confidence = 1.0

    for b in buildings:
        min_x, min_y, max_x, max_y = b.bbox()
        # bbox 가 광선 도달 범위 밖이면 스킵(빠른 거리 하한 프리필터)
        ddx = max(min_x - ox, 0.0, ox - max_x)
        ddy = max(min_y - oy, 0.0, oy - max_y)
        if math.hypot(ddx, ddy) > max_distance_m:
            continue

        if _point_in_ring(ox, oy, b.ring_xy):
            return ShadeResult(
                shaded=True,
                reason="inside_building",
                confidence=1.0 if not b.height_estimated else 0.7,
                blocker_id=b.osm_id,
                blocker_distance_m=0.0,
            )

        d = _ray_ring_min_distance(ox, oy, dx, dy, b.ring_xy)
        if d is None or d > max_distance_m:
            continue

        required = d * tan_alt  # 이 거리에서 태양을 가리려면 필요한 높이
        if b.height_m >= required:
            if nearest_blocker_dist is None or d < nearest_blocker_dist:
                nearest_blocker_dist = d
                nearest_blocker_id = b.osm_id
            # 경계(높이 ≈ 임계) 근접 + 추정 높이면 신뢰도 하락
            if required > 0 and b.height_estimated and b.height_m < required * 1.2:
                confidence = min(confidence, 0.6)

    if nearest_blocker_dist is not None:
        return ShadeResult(
            shaded=True,
            reason="building",
            confidence=confidence,
            blocker_id=nearest_blocker_id,
            blocker_distance_m=nearest_blocker_dist,
        )

    return ShadeResult(shaded=False, reason="sunny", confidence=confidence)
